src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -174,28 +176,39 @@ public override int GetHashCode()
             CompilerTypeSystemContext context,
             CompilationModuleGroup compilationModuleGroup,
             NameMangler nameMangler,
-            ModuleTokenResolver moduleTokenResolver,
-            SignatureContext signatureContext,
             CopiedCorHeaderNode corHeaderNode,
             DebugDirectoryNode debugDirectoryNode,
             ResourceData win32Resources,
             AttributePresenceFilterNode attributePresenceFilterNode,
-            HeaderNode headerNode)
+            ReadyToRunFlags flags)
         {
             TypeSystemContext = context;
             CompilationModuleGroup = compilationModuleGroup;
             Target = context.Target;
             NameMangler = nameMangler;
             MetadataManager = new ReadyToRunTableManager(context);
-            Resolver = moduleTokenResolver;
-            SignatureContext = signatureContext;
             CopiedCorHeaderNode = corHeaderNode;
             DebugDirectoryNode = debugDirectoryNode;
             AttributePresenceFilter = attributePresenceFilterNode;
-            Header = headerNode;
+            Resolver = new ModuleTokenResolver(compilationModuleGroup, TypeSystemContext);
+            Header = new GlobalHeaderNode(Target, flags);
             if (!win32Resources.IsEmpty)
                 Win32ResourcesNode = new Win32ResourcesNode(win32Resources);
 
+            if (CompilationModuleGroup.IsCompositeBuildMode)
+            {
+                // Create a null top-level signature context to force producing module overrides for all signaturess
+                SignatureContext = new SignatureContext(null, Resolver);
+            }
+            else
+            {
+                if (CompilationModuleGroup.CompilationModuleSet.Skip(1).Any())
+                {
+                    throw new InternalCompilerErrorException($"Multiple input modules ({CompilationModuleGroup.CompilationModuleSet.Count()}) are only supported in composite R2R build mode");
```

Comment:
Nit: Make this an assert and have the real check when we're setting up compilation and can write an error out to the user?

Reply:
I have changed the LINQ expression to use <code>Single()</code> to throw when the number of elements is not 1.